src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi

```diff
@@ -112,3 +112,57 @@ def schedule_coro_threadsafe(object coro, object loop):
             )
         else:
             raise
+
+
+def async_generator_to_generator(object agen, object loop):
+    """Converts an async generator into generator."""
+    try:
+        while True:
+            future = asyncio.run_coroutine_threadsafe(
+                agen.__anext__(),
+                loop
+            )
+            response = future.result()
+            if response is EOF:
+                break
+            else:
+                yield response
+    except StopAsyncIteration:
+        # If StopAsyncIteration is raised, end this generator.
+        pass
+
+
+async def generator_to_async_generator(object gen, object loop, object thread_pool):
+    """Converts a generator into async generator.
+
+    The generator might block, so we need to delegate the iteration to thread
+    pool. Also, we can't simply delegate __next__ to the thread pool, otherwise
+    we will see following error:
+
+        TypeError: StopIteration interacts badly with generators and cannot be
+            raised into a Future
+    """
+    queue = asyncio.Queue(loop=loop)
+
+    def yield_to_queue():
+        try:
+            for item in gen:
+                # For an infinite sized queue, the put_nowait should always success
+                loop.call_soon_threadsafe(queue.put_nowait, item)
```

Comment:
Question, what's the difference in the proposal and by just creating a list in memory with all of the items that the generator can provide? I would say, unless Im missing something, that there are no significant differences.

So I would say that using an infinite queue and using the `putt_nowait` we are not doing any kind of cooperation.

Should we try to achieve some kind of cooperation? so only producing messages if and only if messages are being consumed, which if I'm not wrong would be the behavior of two chained iterators.

Reply:
I agree with Pau. If I'm reading this correctly, this `yield_to_queue` function will consume the `gen` generator as fast as it can without ever blocking or yielding. This method is already running in a thread pool. Can we not give the queue size 1 and use a blocking put?

Reply:
Iterators might block between each yield. For example, imagine there is chat application using streaming RPC to communicate. The client might send arbitrary number of messages to the backend. If we try to convert the generator to list, it deadlocks.

Reply:
Updated, now the generator syncs with the event loop.

Sorry, I misunderstood the meaning of `cooperation`. As @gnossen pointed out, this is a valid issue with this design that it might load too many messages.